acme/api/middleware.go

```diff
@@ -0,0 +1,308 @@
+		}
+		if len(jws.Signatures) == 0 {
+			api.WriteError(w, acme.MalformedErr(errors.Errorf("request body does not contain a signature")))
+			return
+		}
+		if len(jws.Signatures) > 1 {
+			api.WriteError(w, acme.MalformedErr(errors.Errorf("request body contains more than one signature")))
+			return
+		}
+
+		sig := jws.Signatures[0]
+		uh := sig.Unprotected
+		if len(uh.KeyID) > 0 ||
+			uh.JSONWebKey != nil ||
+			len(uh.Algorithm) > 0 ||
+			len(uh.Nonce) > 0 ||
+			len(uh.ExtraHeaders) > 0 {
+			api.WriteError(w, acme.MalformedErr(errors.Errorf("unprotected header must not be used")))
+			return
+		}
+		hdr := sig.Protected
+		switch hdr.Algorithm {
+		case jose.RS256, jose.RS384, jose.RS512, jose.ES256, jose.ES384, jose.ES512, jose.EdDSA:
+			// we good
+		default:
+			api.WriteError(w, acme.MalformedErr(errors.Errorf("unsuitable algorithm: %s", hdr.Algorithm)))
+			return
+		}
+
+		// Check the validity/freshness of the Nonce.
+		if err := h.Auth.UseNonce(hdr.Nonce); err != nil {
+			api.WriteError(w, err)
+			return
+		}
+
+		// Check that the JWS url matches the requested url.
+		jwsURL, ok := hdr.ExtraHeaders["url"].(string)
+		if !ok {
+			api.WriteError(w, acme.MalformedErr(errors.Errorf("jws missing url protected header")))
+			return
+		}
+		reqURL := &url.URL{Scheme: "https", Host: r.Host, Path: r.URL.Path}
+		if jwsURL != reqURL.String() {
+			api.WriteError(w, acme.MalformedErr(errors.Errorf("url header in JWS (%s) does not match request url (%s)", jwsURL, reqURL)))
+			return
+		}
+
+		if hdr.JSONWebKey != nil && len(hdr.KeyID) > 0 {
+			api.WriteError(w, acme.MalformedErr(errors.Errorf("jwk and kid are mutually exclusive")))
+			return
+		}
+		if hdr.JSONWebKey == nil && len(hdr.KeyID) == 0 {
+			api.WriteError(w, acme.MalformedErr(errors.Errorf("either jwk or kid must be defined in jws protected header")))
+			return
+		}
+		next(w, r)
+		return
+	}
+}
+
+// extractJWK is a middleware that extracts the JWK from the JWS and saves it
+// in the context. Make sure to parse and validate the JWS before running this
+// middleware.
+func (h *Handler) extractJWK(next nextHTTP) nextHTTP {
+	return func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		jws, ok := jwsFromContext(r)
+		if !ok || jws == nil {
+			api.WriteError(w, acme.ServerInternalErr(errors.Errorf("jws expected in request context")))
+			return
+		}
+		jwk := jws.Signatures[0].Protected.JSONWebKey
+		if jwk == nil {
+			api.WriteError(w, acme.MalformedErr(errors.Errorf("jwk expected in protected header")))
+			return
+		}
+		ctx = context.WithValue(ctx, jwkContextKey, jwk)
+
+		if len(jwk.KeyID) == 0 {
+			kid, err := jwk.Thumbprint(crypto.SHA256)
+			if err != nil {
+				api.WriteError(w, acme.ServerInternalErr(errors.Wrap(err, "error generating jwk thumbprint")))
+				return
+			}
+			jwk.KeyID = base64.RawURLEncoding.EncodeToString(kid)
+		}
+		acc, err := h.Auth.GetAccountByKeyID(jwk.KeyID)
+		switch {
+		case nosql.IsErrNotFound(err):
+			break
```

Comment:
Is this expected?

Reply:
yea. When you're making a NewAccount request there will be a jwk in the jws header which you will later use when creating the new account (so you won't be able to look the account up by kid b/c it doesn't exist yet). This is the only case in which `context.acc == nil` is not an error.

I'll leave a comment in the code.